packages/operator/pkg/utils/model_registry/objectstorage/registry.go

```diff
@@ -0,0 +1,200 @@
+		log.Error(err, "repository creation")
+		return err
+	}
+
+	fullModelPath := path.Join(storage.RemoteConfig.Path, relPath)
+
+	remoteFileIsArchive := strings.HasSuffix(relPath,".zip") || strings.HasSuffix(relPath,".tar.gz")
+
+	// Automatically unzip archive to target location directory
+	if remoteFileIsArchive {
+
+		file, err := ioutil.TempFile("", "*-" + path.Base(relPath))
+		if err != nil {
+			log.Error(err, "Unable to create temp model archive file to fetch into")
+			return err
+		}
+		tempZip := file.Name()
+		if err := storage.Download(tempZip, fullModelPath); err != nil {
+			return err
+		}
+
+		// Ensure that target directory for utils.Unzip destination exists
+		if err = os.MkdirAll(localPath, 0777); err != nil {
+			log.Error(err, "output dir creation")
+			return err
+		}
+		// Unzip archive
+		if err := utils.Unzip(tempZip, localPath); err != nil {
+			log.Error(err, "unzip training data")
+
+			return err
+		}
+
+	} else if err := storage.Download(localPath, fullModelPath); err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+type ConnGetter interface {
+	GetConnection(id string) (*connection.Connection, error)
+}
+
+type ModelRegistry struct {
+	connGetter ConnGetter
+}
+
+
+func NewModelRegistry(cg ConnGetter) *ModelRegistry {
+	return &ModelRegistry{connGetter: cg}
+}
+
+// SyncModel fetch model files from registry to localPath. If localPath is nil
+// then SyncModel set localPath to temp directory and fetch model files there
+// Model can be whether directory or .tar.gz / .zip archive
+// If localPath == "" then temp directory will be created and returned in newLocalPath
+// otherwise newLocalPath == localPath
+func (mr ModelRegistry) SyncModel(connName string, relPath string, localPath string) (newLocalPath string, err error) {
+
+	newLocalPath = localPath
+
+	conn, err := mr.connGetter.GetConnection(connName)
+	if err != nil {
+		return newLocalPath, err
+	}
+
+	if err = conn.DecodeBase64Fields(); err != nil {
+		return newLocalPath, err
+	}
+
+	if localPath == "" {
+		tempDir, err := ioutil.TempDir("", "model-files")
+		if err != nil {
+			return newLocalPath, err
+		}
+		newLocalPath = tempDir
+		defer func() {_ = os.Remove(tempDir)}()
+	}
+
+	if !strings.HasSuffix(newLocalPath, "/") {
+		zap.S().Infof("Adding suffix '/' to %s to mark it as directory " +
+			"(model can be synced only to directory)", newLocalPath)
+		newLocalPath += "/"
+	}
+
+	return newLocalPath, SyncArchiveOrDir(conn.Spec, relPath, newLocalPath)
+}
+
+func getModelNameVersion(modelDir string) (string, string, error) {
+
+	type Model struct {
+		Name string `json:"name"`
+		Version string `json:"version"`
+	}
+
+	type ModelFile struct {
+		Model Model `json:"model"`
+	}
+
+	modelFiles := []string{"odahuflow.project.yaml", "odahuflow.project.yml"}
+
+	items, err := ioutil.ReadDir(modelDir)
+	if err != nil {
+		return "", "", err
+
+	}
+	for _, item := range items {
+		file := item.Name()
+		for _, mFile := range modelFiles {
+			if mFile == file {
```

Comment:
_Code style:_ I'd suggest to reduce nesting:
```
if mFile != file {
    continue
}
// Here goes the rest of code 
```

Reply:
Agree